shotover/src/codec/opensearch.rs

```diff
@@ -0,0 +1,332 @@
+use super::{CodecBuilder, CodecReadError, CodecWriteError, Direction};
+use crate::frame::{
+    opensearch::{HttpHead, RequestParts, ResponseParts},
+    Frame, MessageType, OpenSearchFrame,
+};
+use crate::message::{Encodable, Message, Messages};
+use anyhow::{anyhow, Result};
+use bytes::{Buf, BytesMut};
+use http::{header, HeaderName, HeaderValue, Request, Response};
+use tokio_util::codec::{Decoder, Encoder};
+
+#[derive(Clone)]
+pub struct OpenSearchCodecBuilder {
+    direction: Direction,
+}
+
+impl CodecBuilder for OpenSearchCodecBuilder {
+    type Decoder = OpenSearchDecoder;
+    type Encoder = OpenSearchEncoder;
+
+    fn new(direction: Direction) -> Self {
+        Self { direction }
+    }
+
+    fn build(&self) -> (OpenSearchDecoder, OpenSearchEncoder) {
+        (
+            OpenSearchDecoder::new(self.direction),
+            OpenSearchEncoder::new(self.direction),
+        )
+    }
+
+    fn websocket_subprotocol(&self) -> &'static str {
+        "opensearch"
+    }
+}
+
+pub struct OpenSearchDecoder {
+    direction: Direction,
+    state: State,
+}
+
+impl OpenSearchDecoder {
+    pub fn new(direction: Direction) -> Self {
+        Self {
+            direction,
+            state: State::ParsingResponse,
+        }
+    }
+
+    fn decode_request(&self, src: &mut BytesMut) -> Result<Option<(usize, HttpHead, usize)>> {
```

Comment:
This should return a struct, its not clear what all the fields of this tuple mean.